app/crud/geostore.py

```diff
@@ -115,3 +132,253 @@ async def create_user_area(geometry: Geometry) -> Geostore:
+        admin_provider, admin_version
+    )
+    dataset, version = dv
+
+    src_table: Table = db.table(version)
+    src_table.schema = dataset
+
+    # What exactly is served-up by RW? It looks like it INTENDS to just
+    # serve admin 0s, but the response contains much more
+    where_clause: TextClause = db.text("adm_level=:adm_level").bindparams(adm_level="0")
+
+    gadm_admin_list_columns: List[Column] = [
+        db.column("adm_level"),
+        db.column("gfw_geostore_id"),
+        db.column("gid_0"),
+        db.column("country"),
+    ]
+    sql: Select = (
+        db.select(gadm_admin_list_columns)
+        .select_from(src_table)
+        .where(where_clause)
+        .order_by("gid_0")
+    )
+
+    rows = await get_all_rows(sql)
+
+    return AdminListResponse.parse_obj(
+        {
+            "data": [
+                {
+                    "geostoreId": str(row.gfw_geostore_id),
+                    "iso": str(row.gid_0),
+                    "name": str(row.country),
+                }
+                for row in rows
+            ],
+        }
+    )
+
+
+async def get_all_rows(sql: Select):
+    rows = await db.all(sql)
+
+    return rows
+
+
+async def get_first_row(sql: Select):
+    row = await db.first(sql)
+
+    return row
+
+
+async def get_gadm_geostore(
+    admin_provider: str,
+    admin_version: str,
+    adm_level: int,
+    simplify: float | None,
+    country_id: str,
+    region_id: str | None = None,
+    subregion_id: str | None = None,
+) -> AdminGeostoreResponse:
+    dv: Tuple[str, str] = await admin_params_to_dataset_version(
+        admin_provider, admin_version
+    )
+    dataset, version = dv
+
+    src_table: Table = db.table(version)
+    src_table.schema = dataset
+
+    columns_etc: List[Column | Label] = [
+        db.column("adm_level"),
+        db.column("gfw_area__ha"),
+        db.column("gfw_bbox"),
+        db.column("gfw_geostore_id"),
+        label("level_id", db.column(f"gid_{adm_level}")),
+    ]
+
+    if adm_level == 0:
+        columns_etc.append(label("name", db.column("country")))
+    else:
+        columns_etc.append(label("name", db.column(f"name_{adm_level}")))
+
+    if simplify is None:
+        columns_etc.append(label("geojson", func.ST_AsGeoJSON(db.column("geom"))))
+    else:
+        columns_etc.append(
+            label(
+                "geojson",
+                func.ST_AsGeoJSON(func.ST_Simplify(db.column("geom"), simplify)),
+            )
+        )
+
+    where_clauses: List[TextClause] = [
+        db.text("adm_level=:adm_level").bindparams(adm_level=str(adm_level))
+    ]
+
+    # gid_0 is just a three-character value, but all more specific ids are
+    # followed by an underscore (which has to be escaped because normally in
+    # SQL an underscore is a wildcard) and a revision number (for which we
+    # use an UN-escaped underscore).
+    level_id_pattern: str = country_id
+
+    if adm_level == 0:  # Special-case to avoid slow LIKE
```

Comment:
not part of this PR per se, but should we make sure to add the B-tree soon so LIKE queries are fast?

Reply:
I imagine we should, though I'm not 100% sure a B-tree index is the thing to do. I see suggestions that it can work if we are matching a prefix (a beginning of a pattern), which we are, but it sounds like maybe only if we use the C locale? I'm a little unsure, and also unsure of the ramifications of that (but suspect it won't play nice with our Unicode strings). And then there are these mentions of a trigram index. Here's where I was reading: https://dba.stackexchange.com/a/10696

Reply:
Thanks for creating a ticket for this, @danscales !